### DISCUSSION TITLE
Support Page Animations / Transitions on Android

### DISCUSSION THREAD

**Discussion by shovel-kun** [callstack/react-native-bottom-tabs#209] [Ideas]
# Problem
By default, `BottomNavigationView`should animate when navigating. The default animations [can be found here](https://android.googlesource.com/platform/frameworks/support/+/refs/heads/androidx-main/navigation/navigation-ui/src/main/res/anim), and their [default durations here](https://android.googlesource.com/platform/frameworks/base/+/refs/heads/master/core/res/res/values/config.xml).

[This GitHub issue](https://github.com/android/architecture-components-samples/issues/537#issuecomment-823674468) showed me how to configure the page animation of `BottomNavigationView`. Configuring `res/animator/`in a Native Android app works, but not with `react-native-bottom-tabs`. Presumably, this is because `react-native-screens` handles navigation and their animations.

# Proposed Solution
The way `react-native-screens` handles [stack animation on navigation can be found here](https://github.com/software-mansion/react-native-screens/blob/dd8795f46b310af0aa1c5b5f9895a8807cd66ca2/android/src/main/java/com/swmansion/rnscreens/ScreenStack.kt#L146C1-L244C14). It's technically native, but they are hand-rolling their own navigation, which is why [there are some inconsistencies from native default](https://github.com/software-mansion/react-native-screens/pull/2413). Nonetheless, if we want to go the full native approach like SwiftUI UIView (`@react-navigation/bottom-tabs` [uses RN Animated](https://github.com/react-navigation/react-navigation/blob/main/packages/bottom-tabs/src/views/BottomTabView.tsx#L206]) for page animations), I think this is the place to start.

Currently, I'm not sure how `react-native-bottom-tabs` can tap into `react-native-screen`'s animations. A look into how `react-navigation` implements [Native Stack](https://github.com/react-navigation/react-navigation/blob/main/packages/native-stack/src/views/NativeStackView.native.tsx) might be helpful. Though it seems like the API is unstable at the moment.

**Comment by okwasniewski** [comment]
Hey, 

Thanks for opening the issue. 

> By default, BottomNavigationViewshould animate when navigating. The default animations [can be found here](https://android.googlesource.com/platform/frameworks/support/+/refs/heads/androidx-main/navigation/navigation-ui/src/main/res/anim), and their [default durations here](https://android.googlesource.com/platform/frameworks/base/+/refs/heads/master/core/res/res/values/config.xml).

I'm not sure if this is true. Android views animate when using "stack" navigation, but tabs don't have any animation defined. 

For example, in the Clock app in Android 14, they just immediately change. Can you point me to some Material design guidelines that describe this or show a _native_ Android app that does this (without custom logic)? 

I'm not saying no to adding animations but it would be good to check what's the actual native behaviour. Also its worth noting that BottomNavigationView is responsible only for rendering the bottom bar, it doesn't encapsulate and manage views as SwiftUI's tab view does.

**Comment by shovel-kun** [comment]
Hi, thanks for asking for clarification. I'm new to Open Source and software development in general, so it really helps to be asked these questions.

---

> Can you point me to some Material design guidelines that describe this

[In Navigation bar page > Guidelines > Behavior](https://m3.material.io/components/navigation-bar/guidelines#bff3232e-2a98-4571-ae6a-f4bb4ad413d6), the docs state:
> When you select a navigation bar item (one that’s not currently selected), the app navigates to that destination’s screen using a [top level](https://m3.material.io/m3/pages/transitions/transition-patterns) transition pattern.

Unfortunately, the link they provided is broken lol, but its supposed to link to [Top Level header under Transitions Patterns page](https://m3.material.io/styles/motion/transitions/transition-patterns#f852afd2-396f-49fd-a265-5f6d96680e16), which states:

> This pattern is used to navigate between top-level destinations of an app, like tapping a destination in a Navigation bar.
>
>    Commonly used with: **Navigation bar**, navigation rail, and navigation drawer
>    Read more: [Guidelines](https://m3.material.io/m3/pages/transitions/applying-transitions#048ea44b-1325-4e36-a1c9-8a29fe0c20ea), [Specs](https://m3.material.io/m3/pages/transitions/specs#2833952d-b0c6-43fc-b541-a9f57f4d64b0), [Android implementation](https://github.com/material-components/material-components-android/blob/master/docs/theming/Motion.md#fade-through)

> **The exiting screen quickly fades out and then the entering screen fades in**. Since the content of top level destinations isn't necessarily related, the motion intentionally does not use grouping or persistent elements to create a strong relationship between screens.

That is what is described by the M3 docs.

> or show a native Android app that does this (without custom logic)

You can prove this transition animations are default behavior for `BottomNavigationView` by creating a native Android app with `Bottom Navigation Views Activity` preset in Android Studio:
1. Open Android Studio
2. Click on `New Project` (Top Right)
3. `Templates > Phone and Tablet > Bottom Navigation Views Activity`
4. Build and run the project
5. Notice that FadeIn/FadeOut transition animation plays when navigating between tab bars.

Tested on Android Studio version `2024.2.1`, with API 28 and API 35.

With these reasons in mind, I believe that navigating between bottom tabs should play page animations by default.

---

> Also its worth noting that BottomNavigationView is responsible only for rendering the bottom bar, it doesn't encapsulate and manage views as SwiftUI's tab view does.

You're right. I'm not sure what's the best way of handling this. 

From what I understand, [NavigationBarView has the setupWithNavController method](https://android.googlesource.com/platform/frameworks/support/+/refs/heads/androidx-main/navigation/navigation-ui/src/main/java/androidx/navigation/ui/BottomNavigationView.kt), which wraps around [NavigationUI's setupWithNavController](setupWithNavController). It seems like NavigationUI is responsible for handling navigating as well as animation transition logic, which can be seen in `onNavDestinationSelected`.

But since it is `react-navigation` that is actually handling navigation logic, and not `NavigationUI`, the logic for playing transition animations on navigation is lost, and hence why my earlier method of configuring anims in `res/animator/` won't work.

What are your thoughts?

**Comment by okwasniewski** [comment]
Thanks for providing additional details! Indeed when you create a new app from `Bottom Navigation Views Activity` there is a slight fade animation between views (The spec seems to have a different animation tho 😆) but I think this could be added as a JS animation to views on Android (no need to change native components). 

This should be a fairly simple change where we do a small fade animation when the view changes.

**Comment by okwasniewski** [comment]
I'm moving this to discussions because all feature requests will now be under the "Ideas 💡" tab in GitHub discussions.

**Comment by okwasniewski** [comment]
Released in 0.8.0 🎉
